### PR TITLE
[W.I.P] Implement trace, min, max for Phi/GammaTensor

### PR DESCRIPTION
**Pull request by IshanMi**
## Description
Adding trace, min, max for PhiTensor and GammaTensor as part of the ongoing effort to fill out the Data Science API.

Still missing:
- TensorPointer implementations
- Integration tests (pending fix to issue in test_mean of #6634 )

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
Unit tests for now, integration tests to be added once #6634 is fixed

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

**Review by iamtrask — APPROVED** [review]

File: packages/syft/src/syft/core/tensor/ancestors.py [hunk 1]
```diff
@@ -479,12 +495,22 @@ def _private(
...
+            self.replace_abstraction_top(
+                tensor_type=_PhiTensor(),
+                child=self.child,
+                min_vals=min_vals,
+                max_vals=max_vals,
+                data_subjects=data_subjects,  # type: ignore
+            )  # type: ignore
+        else:
+            self.replace_abstraction_top(
+                tensor_type=_GammaTensor(),
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/tensor/ancestors.py:509]: ❤️

File: packages/syft/src/syft/core/tensor/ancestors.py [hunk 2]
```diff
@@ -479,12 +495,22 @@ def _private(
...
             max_vals = lazyrepeatarray(max_vals, self.child.shape)
 
-        self.replace_abstraction_top(
-            tensor_type=_PhiTensor(),
-            child=self.child,
-            min_vals=min_vals,
-            max_vals=max_vals,
-            data_subjects=data_subjects,  # type: ignore
-        )  # type: ignore
+        unique_data_subjects = len(data_subjects.sum())
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/tensor/ancestors.py:498]: Q:  Should we check at the data point level.
for example DataSubjectArray:
<pre><code>
data = np.array([  [1,2] , 
                   [3,4]])
</code></pre>

dsa = [ [ DSA("natsu"), DSA("natsu"],
                DSA("natsu"), DSA("zerek")]]
Since each data point belongs to different entity, do we consider them  as PhiTensor?
The above code would return a Gamma Tensor.

Suppose if the data subjects were 

dsa = [ [ DSA("natsu","zerek"), DSA("zerek","natsu"],
                DSA("natsu"), DSA("zerek")]]

Like in this example , two of the datapoints have more than one data subjects, so this is a GammaTensor?

TLDR: Maybe we should check datasubjectarray to know if any of the data points have more than one datasubjects?

File: packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py [hunk 3]
```diff
@@ -2706,6 +2943,210 @@ def repeat(
...
+                Minimum of a.
+                If axis is None, the result is a scalar value.
+                If axis is given, the result is an array of dimension a.ndim - 1.
+        """
+
+        if where is None:
+            sources = dict()
+            sources[self.id] = self
+            result = np.amin(self.child, axis=axis, keepdims=keepdims, initial=initial)
+            indices = np.unravel_index(self.child.argmin(axis), shape=self.child.shape)
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py:3026]: 👍

File: packages/syft/src/syft/core/tensor/ancestors.py (hunk 2 above)
Comment (rasswanth-s) [inline packages/syft/src/syft/core/tensor/ancestors.py:498]: Q: I think the DSA works at data point level, Do we consider uniqueness based on the data point or the row of the tensor?
Ex: 
Data  = np.array([[1,2,],
                              [3,4]])

data_subjects = np.array( [ [DSA("natsu"), DSA("natsu")] ] ,
                                              [DSA("natsu"),DSA("zerek")]]
Since the data subjects for a data point are all unique, this is a PhiTensor?
but the above code would consider it as a gamma tensor.

Maybe we could check for uniqueness at the data point level?
Gamma Tensor Data Subjects:

DataSubjects: np.array([[DSA("natsu","zerek"), DSA("grey")]
                                            [DSA("mavis") ,DSA("erza")]]

File: packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py [hunk 4]
```diff
@@ -582,6 +582,243 @@ def sum(
...
+                If a has larger dimensions, then an array of sums along diagonals is returned.
+
+        """
+        attr_path_and_name = "syft.core.tensor.tensor.Tensor.trace"
+        data_subjects = np.array(self.data_subjects).trace(*args, **kwargs)  # type: ignore
+
+        result = TensorWrappedGammaTensorPointer(
+            data_subjects=data_subjects,
+            min_vals=lazyrepeatarray(
+                data=self.min_vals.data, shape=data_subjects.shape
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py:622]: Q: Should we not apply the trace operation on the min/max values of the data?

File: packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py [hunk 5]
```diff
@@ -582,6 +582,243 @@ def sum(
...
+        """
+        attr_path_and_name = "syft.core.tensor.tensor.Tensor.min"
+        data_subjects = np.empty(np.array(self.data_subjects).shape).min(
+            *args, **kwargs
+        )
+
+        result = TensorWrappedGammaTensorPointer(
+            data_subjects=data_subjects,
+            min_vals=lazyrepeatarray(
+                data=self.min_vals.data, shape=data_subjects.shape
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py:698]: Q:Same question do we not compute min operation on min/max vals data.

File: packages/syft/src/syft/core/tensor/autodp/phi_tensor.py [hunk 6]
```diff
@@ -653,6 +654,233 @@ def ones_like(
...
+
+        inherit_tags(
+            attr_path_and_name=attr_path_and_name,
+            result=result,
+            self_obj=self,
+            args=[],
+            kwargs={},
+        )
+        dummy_res = np.ones(self.public_shape)
+        result.public_shape = dummy_res.shape
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/tensor/autodp/phi_tensor.py:807]: NIT: We could do data_subject.shape instead of dummy_res.shape.

File: packages/syft/src/syft/core/tensor/autodp/phi_tensor.py [hunk 7]
```diff
@@ -653,6 +654,233 @@ def ones_like(
...
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/tensor/autodp/phi_tensor.py:879]: NIT: We could do data_subjects.shape instead of dummy_res.shape.

File: packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py [hunk 8]
```diff
@@ -582,6 +582,243 @@ def sum(
 
         return result
 
+    def trace(
```
Comment (rasswanth-s) [inline packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py:585]: MAYBE: could we add some test suites for testing the trace,min,max operations at Pointer level ?

File: packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py (hunk 4 above)
Comment (IshanMi) [reply packages/syft/src/syft/core/tensor/autodp/gamma_tensor.py:622]: Great catch! Fixing in latest commit
